### PR TITLE
Revocation and ledger fault tolerance

### PR DESCRIPTION
**Pull request by seniorjoinu**
1. Implement credential revocation
2. Add retry policy for retrieving ledger operations
3. Add correct behavior on missing artifact in ledger
4. Major abstraction improvement
5. Major refactor

### CODE REVIEW

**Comment by voddan** [comment]
What is the change in `03_cordentity_app_LOGO_color.svg`? If nothing significant changed, then it the change should be reverted and removed from the PR

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/PoolUtils.kt [hunk 1]
```diff
@@ -1,37 +1,25 @@
 package com.luxoft.blockchainlab.hyperledger.indy.utils
 
+import org.apache.commons.io.FileUtils
 import org.hyperledger.indy.sdk.IndyException
```
Comment (alkopnin) [inline indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/PoolUtils.kt]: Why did you change that file?
E.g "io.FileUtils" conflicts with Corda dependencies in supply-chain project.

Could you revert all files not related to implementation? We can check and merge them in separated PR.
It also the same regarding unnecessary formatting. It shouldn't be in the same PR as a new feature.

 Please, revert all not related changes back (formatting, files, etc)

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt [hunk 2]
```diff
@@ -103,15 +106,15 @@ data class KeyCorrectnessProof(val c: String, val xzCap: String, val xrCap: List
 data class Claim(
         val schemaId: String,
         val credDefId: String,
-        val revReg: String?,
+        val revReg: RawJsonMap?,
```
Comment (voddan) [inline indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt]: Why revReg is nullable? The type should be more specific  than RawJsonMap

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt [hunk 3]
```diff
@@ -257,7 +260,7 @@ data class ProofRequestCredentials(
 )
 
 @CordaSerializable
-data class ClaimReferenceInfo(val credInfo: ClaimReference, val interval: String?)
+data class ClaimReferenceInfo(val credInfo: ClaimReference, val interval: Interval? = null)
```
[the remark this answers is not in the document]
Comment (sharpest-tools-in-the-shed) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt]: When revocation is disabled, there is no need to pass non-revoked interval, because prover won't prove non-revocation of this credential.

AnoncredsDemoTest covers this case.

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt (hunk 2 above)
Comment (seniorjoinu) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt]: revReg is nullable because it actually can contain null as a value when revocation is disabled. RawJsonMap is used here because revReg contains revocation registry state at the moment a claim was created, which we don't need to bother until we use ledger as a storage.

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (seniorjoinu) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt]: When revocation is disabled, there is no need for a verifier to specify non-revoked interval, because prover won't prove non-revocation anyway.

This case is covered in AnoncredsDemoTest.

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt [hunk 4]
```diff
@@ -0,0 +1,180 @@
...
+    fun retrieveSchema(schemaId: String) = LedgerService.retrieveSchema(did, pool, schemaId)
+    fun retrieveCredentialDefinition(credDefId: String) = LedgerService.retrieveCredentialDefinition(did, pool, credDefId)
+    fun retrieveRevocationRegistryDefinition(revRegDefId: String) = LedgerService.retrieveRevocationRegistryDefinition(did, pool, revRegDefId)
+    fun retrieveRevocationRegistryEntry(revRegId: String, timestamp: Int) = LedgerService.retrieveRevocationRegistryEntry(did, pool, revRegId, timestamp)
+    fun retrieveRevocationRegistryDelta(revRegDefId: String, interval: Interval) = LedgerService.retrieveRevocationRegistryDelta(did, pool, revRegDefId, interval)
+
+    fun addNym(constructAbout: () -> IndyUser.IdentityDetails) = LedgerService.addNym(did, pool, wallet, constructAbout)
+    fun addNym(about: IndyUser.IdentityDetails) = LedgerService.addNym(did, pool, wallet) { about }
+
+    companion object {
```
[the remark this answers is not in the document]
Comment (seniorjoinu) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt]: This is by design. Casting these methods from static enables you to not to use your wallet information (e.g. everyone with a valid DID can retrieve data from ledger - it's public).

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/PoolUtils.kt [hunk 5]
```diff
@@ -42,10 +30,10 @@ object PoolUtils {
...
+                String.format("{\"reqSignature\":{},\"txn\":{\"data\":{\"data\":{\"alias\":\"Node1\",\"blskey\":\"4N8aUNHSgjQVgkpm8nhNEfDf6txHznoYREg9kirmJrkivgL4oSEimFF6nsQ6M41QvhM2Z33nves5vfSn9n1UwNFJBYtWVnHYMATn76vLuL3zU88KyeAYcHfsih3He6UHcXDxcaecHVz6jhCYz1P2UZn2bDVruL5wXpehgBfBaLKm3Ba\",\"client_ip\":\"%s\",\"client_port\":9702,\"node_ip\":\"%s\",\"node_port\":9701,\"services\":[\"VALIDATOR\"]},\"dest\":\"Gw6pDLhcBcoQesN72qfotTgFa7cbuqZpkX3Xo6pLhPhv\"},\"metadata\":{\"from\":\"Th7MpTaRZVRYnPiabds81Y\"},\"type\":\"0\"},\"txnMetadata\":{\"seqNo\":1,\"txnId\":\"fea82e10e894419fe2bea7d96296a6d46f50f93f9eeda954ec461b2ed2950b62\"},\"ver\":\"1\"}", testPoolIp, testPoolIp),
```
[the remark this answers is not in the document]
Comment (seniorjoinu) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/PoolUtils.kt]: Yes, that's a merge error. Sorry.

[the remark this answers is not in the document]
Comment (seniorjoinu) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/PoolUtils.kt]: This realization uses apache commons fileutils to manage files. @alkopnin also did noticed that issue (check his comment), so we need to decide how would we handle this situation.

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/PoolUtils.kt [hunk 6]
```diff
@@ -55,20 +43,26 @@ object PoolUtils {
...
 
-        try {
-            Pool.createPoolLedgerConfig(poolName, createPoolLedgerConfigJSONParameter.toJson()).get()
-        } catch (e: ExecutionException) {
-            if(getRootCause(e) !is PoolLedgerConfigExistsException) throw e
-            // ok
-        }
-
-        return poolName
+    @Throws(IndyException::class, InterruptedException::class, ExecutionException::class, IOException::class)
```
[the remark this answers is not in the document]
Comment (seniorjoinu) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/PoolUtils.kt]: Check @alkopnin 's comment. We need to decide how to handle this situation together.

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/StorageUtils.kt [hunk 7]
```diff
@@ -0,0 +1,26 @@
+package com.luxoft.blockchainlab.hyperledger.indy.utils
+
+import org.apache.commons.io.FileUtils
+import java.io.File
+import java.io.IOException
+
+object StorageUtils {
```
[the remark this answers is not in the document]
Comment (seniorjoinu) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/StorageUtils.kt]: That's a great idea.

File: indy-utils/src/test/kotlin/com/luxoft/blockchainlab/hyperledger/indy/IndyIntegrationTest.kt [hunk 8]
```diff
@@ -1,30 +1,12 @@
 package com.luxoft.blockchainlab.hyperledger.indy
 
-import org.hyperledger.indy.sdk.crypto.CryptoJSONParameters
```
[the remark this answers is not in the document]
Comment (seniorjoinu) [reply indy-utils/src/test/kotlin/com/luxoft/blockchainlab/hyperledger/indy/IndyIntegrationTest.kt]: Okay, let's discuss it.

Anyway, now we don't need this stuff.

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt [hunk 9]
```diff
@@ -0,0 +1,180 @@
...
+import org.hyperledger.indy.sdk.ledger.Ledger
+import org.hyperledger.indy.sdk.pool.Pool
+import org.hyperledger.indy.sdk.wallet.Wallet
+import org.slf4j.LoggerFactory
+import java.util.*
+
+const val DELAY_MS = 2000
+const val RETRY_TIMES = 5
+
+class LedgerService(
```
[the remark this answers is not in the document]
Comment (seniorjoinu) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt]: Agree.

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt [hunk 10]
```diff
@@ -293,7 +331,9 @@ data class RequestedPredicateInfo(
 )
 
 /**
- * proof request json
+ * Represents proof request
+ *
+ * Example:
```
Comment (voddan) [inline indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt]: Только это не Example, а Format тогда уж

**Comment by voddan** [comment]
Я тут подумал над нашим вчерашним разговором, и вынужден сказать что не согласен с решением отложить вопрос с LedgerService на потом. Это относительно быстро сделать, и у нас сейчас нету жесткого дедлайна. Не понимаю зачем допускать неидеальный код в проект только чтобы когда-нибудь потом его исправить.

File: build.gradle [hunk 11]
```diff
@@ -5,6 +5,7 @@ buildscript {
     ext.quasar_version = '0.7.9'
     ext.junit_version = '4.12'
     ext.kotlin_version = '1.2.30'
+    ext.kotlin_coroutines_version = '0.25.0'
```
Comment (seniorjoinu) [inline build.gradle]: ?

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (voddan) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/Models.kt]: Could you write in the docs about it, please?

File: cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateClaimDefinitionFlow.kt [hunk 12]
```diff
@@ -0,0 +1,46 @@
...
+    @StartableByRPC
+    class Authority(private val schemaId: String, private val maxCredNumber: Int = 100) : FlowLogic<CreateClaimDefinitionFlowResult>() {
+
+        @Suspendable
+        override fun call(): CreateClaimDefinitionFlowResult {
+            try {
+                val credDef = indyUser().createClaimDefinition(schemaId, true)
+                val revReg = indyUser().createRevocationRegistry(credDef.id, maxCredNumber)
+
+                return CreateClaimDefinitionFlowResult(credDef.id, revReg.definition.id)
```
[the remark this answers is not in the document]
Comment (alkopnin) [reply cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateClaimDefinitionFlow.kt]: Why do not  create CredentialDef State and don't handle additional id on client side?

File: cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/VerifyClaimFlow.kt [hunk 13]
```diff
@@ -79,12 +83,17 @@ object VerifyClaimFlow {
...
+                        nonRevoked = nonRevoked
+                )
 
                 val verifyClaimOut = flowSession.sendAndReceive<ProofInfo>(proofRequest).unwrap { proof ->
-                    val claimProofOut = IndyClaimProof(identifier, proofRequest, proof, listOf(ourIdentity, prover))
+                    val usedData = indyUser().getDataUsedInProof(proofRequest, proof)
+                    val claimProofOut = IndyClaimProof(identifier, proofRequest, proof, usedData, listOf(ourIdentity, prover))
 
-                    if (!IndyUser.verifyProof(claimProofOut.proofReq, proof)) throw FlowException("Proof verification failed")
+                    if (!indyUser().verifyProof(claimProofOut.proofReq, proof, usedData)) throw FlowException("Proof verification failed")
```
[the remark this answers is not in the document]
Comment (seniorjoinu) [reply cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/VerifyClaimFlow.kt]: Because verifier has his own indyuser object in this case. I can revert it if there is a reason.

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt [hunk 14]
```diff
@@ -0,0 +1,284 @@
...
+         * @param did           indy user did
+         * @param pool          indy pool handle
+         * @param revRegId      revocation registry id
+         * @param timestamp     time from unix epoch in seconds representing time moment you are interested in
+         *                      e.g. if you want to know current revocation state, you pass 'now' as a timestamp
+         *
+         * @return              revocation registry entry or null if none exists on ledger
+         */
+        fun retrieveRevocationRegistryEntry(did: String?, pool: Pool, revRegId: String, timestamp: Int): Pair<Int, RevocationRegistryEntry>? = runBlocking {
+            val result: Pair<Int, RevocationRegistryEntry>? = null
```
[the remark this answers is not in the document]
Comment (alkopnin) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt]: I slightly concerned about both approaches. Let's investigate best practice in Corda. It may be restricted to sleep during the flow execution.

[the remark this answers is not in the document]
Comment (seniorjoinu) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt]: This subject should be investigated more carefully. This function is called from some corda flows, so it shouldn't operate with threads to be correctly serialized with Kryo. Coroutine will work with 100% probability because it's just a function with a context.

[the remark this answers is not in the document]
Comment (voddan) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt]: > Coroutine will work with 100% probability

No, `runBlocking` blocks a thread the same way `Thread.sleep` does. AFAIU Quasar, it is okey to sleep on a thread since any Java function may do this.

File: indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt [hunk 15]
```diff
@@ -0,0 +1,284 @@
...
+import kotlinx.coroutines.experimental.delay
+import kotlinx.coroutines.experimental.runBlocking
+import org.hyperledger.indy.sdk.ledger.Ledger
+import org.hyperledger.indy.sdk.pool.Pool
+import org.hyperledger.indy.sdk.wallet.Wallet
+import org.slf4j.LoggerFactory
+import java.util.*
+
+const val DELAY_MS = 2000
+const val RETRY_TIMES = 5
```
[the remark this answers is not in the document]
Comment (voddan) [reply indy-utils/src/main/kotlin/com/luxoft/blockchainlab/hyperledger/indy/utils/LedgerService.kt]: What I meant was that I need to change the `RETRY_TIMES` when I create a LedgerService

File: cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateClaimDefinitionFlow.kt [hunk 16]
```diff
@@ -0,0 +1,67 @@
...
+    @StartableByRPC
+    class Authority(private val schemaId: String, private val maxCredNumber: Int = 100) : FlowLogic<String>() {
+
+        @Suspendable
+        override fun call(): String {
+            try {
+                val schemaFromVault = getIndySchemaState(schemaId)
+                    ?: throw RuntimeException("There is no schema with id: $schemaId in vault")
+
+                val credDef = indyUser().createClaimDefinition(schemaId, true)
```
Comment (alkopnin) [inline cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateClaimDefinitionFlow.kt]: Need to check that credential definition already exists. If it exist and number of issued credential is less than limit (maxCredNumber) - throw CredentialDefinition already exist, otherwise if limit is rich - throw CredentialDefinition exist and need to be upgraded. If no one - create it.

Upgrade process has to be performed in separated flow - UpgradeCredentialDefinition. (can be implemented in separate PR)

File: cordapp-contracts-states/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/data/state/IndyClaimDefinition.kt [hunk 17]
```diff
@@ -0,0 +1,29 @@
...
+import net.corda.core.schemas.PersistentState
+import net.corda.core.schemas.QueryableState
+
+
+class IndyClaimDefinition(
+    val schemaId: String,
+    val claimDefId: String,
+    val revRegId: String,
+    override val participants: List<AbstractParty>
+) : LinearState, QueryableState {
```
[the remark this answers is not in the document]
Comment (alkopnin) [reply cordapp-contracts-states/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/data/state/IndyClaimDefinition.kt]: The state also has obtain counter of number of issued credentials. This state has to be consumed on every new credential and counter is incremented. Contract need to check these counters.

File: cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateClaimDefinitionFlow.kt [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+                val commandType = ClaimMetadataChecker.Command.Create()
+                val signers = listOf(ourIdentity.owningKey)
+
+                val command = Command(commandType, signers)
+
+                val schemaCmdType = SchemaChecker.Command.Use()
+                val schemaCmd = Command(schemaCmdType, signers)
+
+                val trxBuilder = TransactionBuilder(whoIsNotary())
+                    .withItems(claimDefOut, command, schemaFromVault, schemaCmd)
```
Comment (alkopnin) [inline cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateClaimDefinitionFlow.kt]: Schema state has to be consumed and created new one.

I suggest to use more clear names:
Schema:
schemaIn
schemaOut
consumeSchemaCmd

CredentialDefinition:
credentialDefinitionIn
credentialDefinitionOut
createCredentialDefinitionCmd

File: cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateClaimDefinitionFlow.kt [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+
+                trxBuilder.toWireTransaction(serviceHub)
+                    .toLedgerTransaction(serviceHub)
+                    .verify()
+
+                val selfSignedTx = serviceHub.signInitialTransaction(trxBuilder, ourIdentity.owningKey)
+
+                subFlow(FinalityFlow(selfSignedTx))
+
+                return claimDef.claimDefId
```
Comment (alkopnin) [inline cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateClaimDefinitionFlow.kt]: Im not sure that we need to return it back. 
Schema or CredentialDefinition could be retrieved from the vault

File: cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateSchemaFlow.kt [hunk 20]
```diff
@@ -27,8 +33,28 @@ object CreateSchemaFlow {
...
-                // create new schema
-                return indyUser().createSchema(schemaName, schemaVersion, schemaAttributes).id
+
+                val schema = indyUser().createSchema(schemaName, schemaVersion, schemaAttributes)
+                val indySchema = IndySchema(schema.id, listOf(ourIdentity))
+                val schemaOut = StateAndContract(indySchema, SchemaChecker::class.java.name)
+
+                val commandType = SchemaChecker.Command.Create()
+                val signers = listOf(ourIdentity.owningKey)
+                val command = Command(commandType, signers)
```
Comment (alkopnin) [inline cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateSchemaFlow.kt]: See recommendation for notation above

File: cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateSchemaFlow.kt [hunk 21]
```diff
@@ -27,8 +33,28 @@ object CreateSchemaFlow {
...
+
+                trxBuilder.toWireTransaction(serviceHub)
+                    .toLedgerTransaction(serviceHub)
+                    .verify()
+
+                val selfSignedTx = serviceHub.signInitialTransaction(trxBuilder, ourIdentity.owningKey)
+
+                subFlow(FinalityFlow(selfSignedTx))
+
+                return schema.id
```
Comment (alkopnin) [inline cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/CreateSchemaFlow.kt]: Don't need to return it back. Id should be retrieved from state

File: cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/IssueClaimFlow.kt [hunk 22]
```diff
@@ -69,8 +81,11 @@ object IssueClaimFlow {
 
                 val newClaimCmd = Command(newClaimData, newClaimSigners)
 
+                val claimDefCmdType = ClaimMetadataChecker.Command.Use()
+                val claimDefCmd = Command(claimDefCmdType, newClaimSigners)
+
                 val trxBuilder = TransactionBuilder(whoIsNotary())
-                        .withItems(newClaimOut, newClaimCmd)
+                        .withItems(newClaimOut, newClaimCmd, credDefState, claimDefCmd)
```
Comment (alkopnin) [inline cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/IssueClaimFlow.kt]: CredentialDefinition need to be consumed and produced new one. 
Current number of credential should be increased and checked in the contract.

File: cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/IssueClaimFlow.kt [hunk 23]
```diff
@@ -100,19 +115,24 @@ object IssueClaimFlow {
...
+
+                        outputs.forEach {
+                            val state = it.data
+
+                            when (state) {
+                                is IndyClaim -> {
+                                    require(state.claimRequestInfo == claimRequestInfo) { "Received incorrect ClaimReq" }
+                                    indyUser().receiveClaim(state.claimInfo, state.claimRequestInfo, offer)
+                                }
+                                is IndyClaimDefinition -> println("Got indy claim definition")
```
Comment (alkopnin) [inline cordapp/src/main/kotlin/com.luxoft.blockchainlab.corda.hyperledger.indy/flow/IssueClaimFlow.kt]: don't use println. Lets use logger instead
